### ISSUE TITLE
Migrating Single-Node ClickHouse Instances to Replicated Setup

### ISSUE THREAD

**Issue by honghainguyen777** [ClickHouse/ClickHouse#70118]
### Question

We are upgrading our single-node ClickHouse instances (from v20.8.9.6, 3 nodes with differing data) used by our self-hosted Sentry (K8s). The challenge is migrating to a replicated architecture while preserving data across nodes.

Attempting to migrate each MergeTree table to ReplicatedMergeTree (via `INSERT INTO newTable SELECT * FROM oldTable`) results in data loss due to synchronization deleting differences between nodes.

We also tried keeping the old MergeTree tables alongside new ReplicatedMergeTree tables (created by Snuba). However, this led to data imbalances. Some events are missing when queried, possibly due to uneven distribution across nodes.

To address this, we are considering merging all MergeTree tables from multiple nodes into a single-node setup, followed by a replication setup. **Is there a way to merge these tables without duplicating data**?

**Comment by Felixoid** [comment]
But then your data is sharded. I assume, you have a Distributed table on top of it? That's why I suggested 3 new nodes.

in any case, I am glad it worked

**Comment by honghainguyen777** [comment]
Yes, we have distributed tables on top of the MergeTree tables. I actually used the distributed tables as the source, then created new ReplicatedMergeTree tables, inserted date from the dist tables to those tables.

I have not tried to create new nodes on top of the existing ones (a bit too late after reading your suggestions). But thank you so much for that! I would experiment it in one of our Sentry instance later. It was challenging updating our self-hosted Sentry from v11 to v25 without the loss of our clickhouse data.

Have a good day!
